py/desitarget/mock/build.py

```diff
@@ -1495,52 +446,98 @@ def finish_catalog(targets, truth, skytargets, skytruth, nside, healpix_id, seed
-        subpriority = rand.uniform(0.0, 1.0, size=n_obj+n_sky)
+    nobj = len(targets)
+    nsky = len(skytargets)
+    log.info('Summary: ntargets = {}, nsky = {} in pixel {}.'.format(nobj, nsky, healpix))
 
-        truth['TARGETID'][:] = targetid[:n_obj]
-        targets['TARGETID'][:] = targetid[:n_obj]
-        targets['SUBPRIORITY'][:] = subpriority[:n_obj]
-    
-    if n_sky > 0:
-        skytargets['TARGETID'][:] = targetid[n_obj:]
-        skytargets['SUBPRIORITY'][:] = subpriority[n_obj:]
-        
-    if n_obj > 0:
-        targpix = radec2pix(nside, targets['RA'], targets['DEC'])
-        targets['HPXPIXEL'][:] = targpix
+    if use_brickid:
 
-    if n_sky > 0:
-        targpix = radec2pix(nside, skytargets['RA'], skytargets['DEC'])
-        skytargets['HPXPIXEL'][:] = targpix
-    
+        # Assign the correct BRICKID and unique OBJIDs for every object on this brick.
+        from desiutil.brick import Bricks
+
+        brick_info = Bricks().to_table()
+
+        if nobj > 0 and nsky > 0:
+            allbrickname = set(targets['BRICKNAME']) | set(skytargets['BRICKNAME'])
+        if nobj > 0 and nsky == 0:
+            allbrickname = set(targets['BRICKNAME'])
+        if nobj == 0 and nsky > 0:
+            allbrickname = set(skytargets['BRICKNAME'])
+
+        for brickname in allbrickname:
+            iinfo = np.where(brickname == brick_info['BRICKNAME'])[0]
+
+            nobj_brick = 0
+            if nobj > 0:
+                itarg = np.where(brickname == targets['BRICKNAME'])[0]
+                nobj_brick = len(itarg)
+                targets['BRICKID'][itarg] = brick_info['BRICKID'][iinfo]
+                targets['BRICK_OBJID'][itarg] = np.arange(nobj_brick)
+
+            if nsky > 0:
+                iskytarg = np.where(brickname == skytargets['BRICKNAME'])[0]
+                nsky_brick = len(iskytarg)
+                skytargets['BRICKID'][iskytarg] = brick_info['BRICKID'][iinfo]
+                skytargets['BRICK_OBJID'][iskytarg] = np.arange(nsky_brick) + nobj_brick
+
+        subpriority = rand.uniform(0.0, 1.0, size=nobj+nsky)
+
+        if nobj > 0:
+            targetid = encode_targetid(objid=targets['BRICK_OBJID'], brickid=targets['BRICKID'], mock=1)
+            truth['TARGETID'][:] = targetid
+            targets['TARGETID'][:] = targetid
+            targets['SUBPRIORITY'][:] = subpriority[:nobj]
+
+            targets['HPXPIXEL'][:] = radec2pix(nside, targets['RA'], targets['DEC'])
+
+        if nsky > 0:
+            skytargetid = encode_targetid(objid=skytargets['BRICK_OBJID'], brickid=skytargets['BRICKID'], mock=1, sky=1)
+            skytargets['TARGETID'][:] = skytargetid
+            skytargets['SUBPRIORITY'][:] = subpriority[nobj:]
+
+            skytargets['HPXPIXEL'][:] = radec2pix(nside, skytargets['RA'], skytargets['DEC'])
+                
+    else:
+        objid = np.arange(nobj + nsky)
+        targetid = encode_targetid(objid=objid, brickid=healpix, mock=1)
```

Comment:
Although your uses of `encode_targetid` are formally correct (thanks for making this switch!) this line is potentially mildly dangerous. `brickid` is set to a maximum of 20 bits, which will be exceeded at `nside = 512` (but is safe for `nside < 512`). I doubt that we'll ever use `nside=512` and `encode_targetid` is smart enough to flag an error if the bit-allowance of 20 bits for `brickid` is exceeded, but you might want to capture an error for the `nside > 256` case, here?

Reply:
Thanks @geordie666 I've added a ValueError exception if nside > 256 at the start of `mock.build.targets_truth`.